Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs

```diff
@@ -0,0 +1,593 @@
+        };
+
+        private static readonly HashSet<string> ExcludedYamlAssetExtensions = new HashSet<string> { ".jpg", ".csv", ".meta", ".pfx", ".txt", ".nuspec", ".asmdef", ".yml", ".cs", ".md", ".json", ".ttf", ".png", ".shader", ".wav", ".bin", ".gltf", ".glb", ".fbx", ".FBX", ".pdf", ".cginc" };
+        private static readonly HashSet<string> ExcludedSuffixFromCopy = new HashSet<string>() { ".cs", ".cs.meta", ".asmdef", ".asmdef.meta" };
+
+        private static Dictionary<string, string> nonClassDictionary = new Dictionary<string, string>(); //Guid, FileName
+
+        private const string ScriptFileIdConstant = "11500000";
+
+        [MenuItem("Assets/Retarget To DLL")]
+        public static void RetargetAssets()
+        {
+            try
+            {
+                Debug.Log("Starting to Retarget Assets.");
+                RunRetargetToDLL();
+                Debug.Log("Complete.");
+            }
+            catch (Exception ex)
+            {
+                Debug.LogError("Failed.");
+                Debug.LogException(ex);
+            }
+        }
+
+        private static void RunRetargetToDLL()
+        {
+            string[] allFilesUnderAssets = Directory.GetFiles(Application.dataPath, "*", SearchOption.AllDirectories);
+
+            Dictionary<string, ClassInformation> scriptFilesReferences = ProcessScripts(allFilesUnderAssets);
+            Debug.Log($"Found {scriptFilesReferences.Count} script file references.");
+
+            // DLL name to Guid
+            Dictionary<string, string> asmDefMappings = RetrieveAsmDefGuids(allFilesUnderAssets);
+
+            Dictionary<string, ClassInformation> compiledClassReferences = ProcessCompiledDLLs("PackagedAssemblies", Application.dataPath.Replace("Assets", "NuGet/Plugins/EditorPlayer"), asmDefMappings);
+            Debug.Log($"Found {compiledClassReferences.Count} compiled class references.");
+
+            Dictionary<string, Tuple<string, long>> remapDictionary = new Dictionary<string, Tuple<string, long>>();
+
+            foreach (var pair in scriptFilesReferences)
+            {
+                if (compiledClassReferences.TryGetValue(pair.Key, out ClassInformation compiledClassInfo))
+                {
+                    remapDictionary.Add(pair.Value.Guid, new Tuple<string, long>(compiledClassInfo.Guid, compiledClassInfo.FileId));
+                }
+                else
+                {
+                    // Switch to throwing exception later
+                    Debug.LogWarning($"Can't find a compiled version of the script: {pair.Key}; guid: {pair.Value.Guid}");
+                }
+            }
+
+            ProcessYAMLAssets(allFilesUnderAssets, Application.dataPath.Replace("Assets", "NuGet/Content"), remapDictionary, asmDefMappings);
+        }
+
+        private static Dictionary<string, string> RetrieveAsmDefGuids(string[] allFiles)
+        {
+            int lengthOfPrefix = Application.dataPath.IndexOf("Assets");
```

Comment:
>int lengthOfPrefix = Application.dataPath.IndexOf("Assets"); [](start = 12, length = 60)

Super nit: This is shared between a couple of places, consider having a helper function/constant so that we don't have to have multiple instances of hardcoded "Assets" strings.

Reply:
Yeah I know, but it's a minor in terms of time, and I would rather keep it here for easier code understanding/readability as to how the value is gotten.

---
In reply to: [277826520](https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/4051#discussion_r277826520) [](ancestors = 277826520)